Joss paper (#104)
CMakePP/CMinx: paper.md

seamlessly into existing CMake build systems. CMinx's output is highly
customizable and easily controlled via a YAML [@yaml] configuration file.
CMinx has already proved to be an invaluable
productivity tool in the authors' other projects, and, given that other
scientific software projects also rely heavily on source code written in CMake,
we anticipate CMinx will prove invaluable to many additional projects as well.

Concurrent with the submission of this manuscript, we have also released the
first production version of CMinx, version 1.0.0. CMinx can be obtained
from the Python Packaging Index via "`pip install CMinx`". Alternatively, CMinx
can be used as a CMake module via CMake's "`FetchContent`" command.
Despite only just releasing 1.0.0, the CMinx GitHub organization
has already started to see attention and interest from developers not
affiliated with the authors. So while CMinx might not be considered research
software, we anticipate CMinx to be a useful productivity tool for the large
swath of research software that uses CMake as their build system.

# Statement of need

The process of building a software package written in a compiled language

> (*e.g.*, C or C++) has always been complicated. Over the years, various build

Reviewer: ```suggestion
(*e.g.*, C, C++ or Fortran) has always been complicated. Over the years, various build
```
I'd add Fortran as well, as that adds its own intricacies to the build system.

system solutions have evolved to ease the process. Historically, there
has been a propensity to treat each build system as a one-off use case. This is
understandable since build systems have tended to be relatively small
and tightly coupled to the structure and purpose of the package.

> With build system complexity at an all-time high[@snir2014,@xSDK], there is an

Reviewer: ```suggestion
With build system complexity at an all-time high [@snir2014,@xSDK], there is an
```
should there be a space here?

increasing need to treat the underlying build system infrastructure as code.
This means that the build system should be modularized, and those modules
should be documented, tested, and reusable. With the popularity of C/C++ for
high-performance computing, "build system" is increasingly becoming synonymous

> with CMake. Therefore with the advent of the exascale era, there is a desperate

Reviewer: ```suggestion
with CMake: there is a desperate
```
I think the need is there whether you're application targets the exascale or not.

Concurrent with the submission of this manuscript, we have also released the
first production version of CMinx, version 1.0.0. CMinx can be obtained
from the Python Packaging Index via "`pip install CMinx`". Alternatively, CMinx
can be used as a CMake module via CMake's "`FetchContent`" command.
Despite only just releasing 1.0.0, the CMinx GitHub organization
has already started to see attention and interest from developers not
affiliated with the authors. So while CMinx might not be considered research
software, we anticipate CMinx to be a useful productivity tool for the large
swath of research software that uses CMake as their build system.

# Statement of need

The process of building a software package written in a compiled language
(*e.g.*, C, C++ or Fortran) has always been complicated. Over the years, various build
system solutions have evolved to ease the process. Historically, there
has been a propensity to treat each build system as a one-off use case. This is
understandable since build systems have tended to be relatively small
and tightly coupled to the structure and purpose of the package.

> With build system complexity at an all-time high [@snir2014,@xSDK], there is an

Reviewer 2: Needs semicolon to render properly
```suggestion
With build system complexity at an all-time high [@snir2014; @xSDK], there is an
```

---

> title: 'CMinx: Restructured Text Documentation Generator for CMake'

Reviewer 3: I would say the most important part of the title is that it's for CMake, so it would probably be best if we could reword it to move CMake closer to the beginning of the title. Maybe "CMinx: The CMake Documentation Generator"? Or "CMinx: Documenting CMake with Restructured Text"?

With the current title, readers might miss the CMake part and assume it's an alternative to already existing RST documentation generators

API documentation as part of a normal Sphinx workflow. This Sphinx plugin 
is distributed with the source code for the CMake interpreter, and is 
also available in a GitHub repository mirror [@sphinx_plugin]. 
For completeness, we note that a couple of similar Sphinx
plugins [@official_sphinx_domain; @marco_koch] also exist, but they appear
to have been abandoned.

To our knowledge, all of the aforementioned documentation solutions simply 
extract the reST API documentation verbatim. Notably this means the
developer is responsible for manually: adding the function/macro signatures 
to the documentation, listing the function's parameters, and the overall
formatting (*e.g.*, titles and subtitles, creation of parameter 
tables, underline/overline consistency). For build system developers 
maintaining one or two CMake modules, these are admittedly pretty minor 
inconveniences; however, for an organizations which maintaining a substantial 
CMake code base, these "minor inconveniences" can impact productivity. 
In order to increase the productivity of the CMakePP team, we created CMinx.

Concurrent with the submission of this manuscript, we have also released the 

> first production version of CMinx. Promisingly, the CMinx GitHub organization

Reviewer 3: > first production version of CMinx

This would be a great place to put a reference to the PyPI page. We should probably also mention that it has a CMake interface somewhere and put a link to the GitHub page there

development of the target software package. For example, CMake's
`find_package` module simplifies dependency management and the `CTest` package
eases the process of testing the resulting software. Additional CMake tools
can be created by writing CMake modules. While the CMake language is
flexible and relatively simple, it is not without its pitfalls. Unfortunately
tools  to facilitate the development of CMake modules are relatively sparse.
Here we introduce CMinx, a tool for generating API documentation for CMake
modules. API documentation is arguably one of the most basic elements of a
software development ecosystem and it is our hope that CMinx will serve as the
foundation for a robust CMake development ecosystem.

Anecdotal evidence [@official_solution] indicates that the developers of the
CMake language (*i.e.*, Kitware) internally write their API documentation
using reST and Sphinx. Following best practices, this reST API documentation
resides next to the CMake code being described. Kitware has also written a
Sphinx plugin that makes it easy to extract the API documentation as part of a
normal Sphinx workflow. This Sphinx plugin is distributed with the source code

> for the CMake interpreter, and is also available in a GitHub repository mirror

Reviewer 3: > This Sphinx plugin is distributed with the source code
for the CMake interpreter,

Comma after "interpreter" should be removed

[@sphinx_plugin]. For completeness, we note that similar Sphinx plugins

> [@official_sphinx_domain; @marco_koch] have been independently developed, but

Reviewer 3: > have been independently developed, but

Comma after "developed" should be removed

appear to now be abandoned.

To our knowledge, all of the aforementioned Sphinx plugins simply extract the

> reST API documentation verbatim. Notably this means the developer is responsible

Reviewer 3: > Notably this means the developer

Should be a comma after "notably"

for manually: adding the function/macro signatures to the documentation, listing
the function's parameters, and the overall formatting (*e.g.*, titles and
subtitles, creation of parameter tables, underline/overline consistency). For
build system developers maintaining one or two CMake modules, these are
admittedly pretty minor inconveniences; however, for organizations
maintaining a substantial CMake code base, these "minor inconveniences" can
impact productivity, particularly when ensuring consistency. CMinx
differs from previous solutions primarily in three ways. First, CMinx
understands the grammar of the CMake language, meaning CMinx can automatically
generate some of the documentation by "reading" the source code. Second, CMinx
generates static reST files; this decreases the number of stub files developers
need to maintain and makes it easier for the resulting documentation to be

> used in workflows which do not rely on Sphinx. Finally, CMinx has a CMake API

Reviewer 3: > used in workflows which do not rely on Sphinx.

"Which" should be swapped for "that"

> used in workflows which do not rely on Sphinx. Finally, CMinx has a CMake API

Reviewer 3: > Finally, CMinx has a CMake API
so that it integrates more easily into existing CMake workflows.

I'm not quite sure why but something doesn't seem quite right about this sentence. Maybe something like "Finally, CMinx has a CMake API to integrate more easily with existing CMake workflows"?

Or perhaps something like "Finally, CMinx has a CMake API to allow easy integration with existing CMake workflows"

---
title: 'CMinx: A CMake Documentation Generator'
tags:
  - Python
  - CMake
  - Documentation
authors:
  - name: Branden Butler
    equal-contrib: true
    affiliation: "1, 2"
  - name: Ryan M. Richard
    orcid: 0000-0003-4235-5179
    equal-contrib: false
    corresponding: true # (This is how to denote the corresponding author)
    affiliation: "1, 2"
affiliations:
 - name: Ames National Laboratory, Ames, IA, USA
   index: 1
 - name: Iowa State University, Ames, IA, USA
   index: 2
date: August 2022
bibliography: paper.bib
---

# Summary

This manuscript introduces CMinx, a program for generating
application programming interface (API) documentation for CMake modules
(code written in the CMake language). Since most of CMinx's intended audience
is comprised of C/C++ developers, CMinx is designed to operate similar to
Doxygen [@doxygen] (the *de facto* C/C++ API documentation tool).
Specifically, developers annotate their CMake source with documentation
comments (traditional CMake block comments, starting with an extra "`[`"

> character). The documentation comments, written in reST [@rest],

Author: I'm inclined to keep this parenthetical since I think it's important to introduce the "documentation comment" term, and define it.

> character). The documentation comments, written in reST [@rest],

Reviewer 2: Parentheses can state something less important than the parent sentence, something that can be skipped. If you want to define the term here, I suggest to make it more clear. How about

> Specifically, developers annotate their CMake source with "documentation comments", which are traditional CMake block comments starting with an extra "`[`" character.

> character). The documentation comments, written in reST [@rest],

Author: I like your suggestion.

describe to the reader how the functions, parameters, and variables should be
used. Running CMinx on the annotated source code generates reST files containing
the API documentation. The reST files can then be converted into static
websites with tools such as Sphinx [@sphinx] or
easily converted to another format via Pandoc [@pandoc].

Unlike other solutions for documenting CMake modules, CMinx knows the CMake
language's grammar. This enables CMinx to automatically extract function/macro

> signatures (even when functions are not documented). CMinx also integrates

Reviewer 2: Avoid parentheses
```suggestion
signatures, even when functions are not documented. CMinx also integrates
```

seamlessly into existing CMake build systems. CMinx's output is highly

> customizable and easily controlled via a YAML [@yaml] configuration file.

Reviewer 2: I can find this claim nowhere in the documentation. How can CMinx be customized with a configuration file? Please clarify or remove the claim.

Author: This is a relatively new feature. At the moment I think it's only documented in the ``cminx --help`` message. I'll make sure that gets into the documentation.

CMinx has already proved to be an invaluable
productivity tool in the authors' other projects, and, given that other
scientific software projects also rely heavily on source code written in CMake,
we anticipate CMinx will prove invaluable to many additional projects as well.

# Statement of need

> 
> The exascale-era of high-performance scientific computing arguably started when
> the Frontier supercomputer achieved a performance of more than 1.1 exaFLOPs
> (10$^{18}$ floating-point operations per second) on the High-Performance
> Linpack Benchmark [@frontier]. With the breakdown of Dennard scaling
> (*i.e.*, the power usage of a transistor is proportional to its area),
> achieving exascale performance is presently only possible by relying
> on heterogeneous hardware. Unfortunately, hardware heterogeneity significantly
> complicates the process of developing high-performance scientific software.
> For many high-performance software packages, these complications start in the
> build system.

Reviewer 2: I find that this paragraph to be superfluous. The issue of maintaining build system code has little to do with exascale computing per se. The actual connection is done in the following paragraph, so I would suggest to remove this paragraph entirely. I think the paper actually works well without it.

```suggestion
```

The process of building a software package written in a compiled language
(*e.g.*, C or C++) has always been complicated. Over the years, various build
system solutions have evolved to ease the process. Historically, there
has been a propensity to treat each build system as a one-off use case. This is

> understandable since build systems have tended to be relatively small (*i.e.*,
> less than ~1K lines of code) and tightly coupled to the identity of the package.

Reviewer 2: Unnecessary specification in parentheses. However, "identity" is too vague in my opinion:
```suggestion
understandable since build systems have tended to be relatively small
and tightly coupled to the structure and purpose of the package.
```

With build system complexity at an all-time high, there is now an
increasing desire (at least within scientific computing communities) to treat
the underlying build system infrastructure as code. This means that the
build system should be modularized, and those modules should be documented,

> tested, and reusable. With the popularity of C/C++ for high-performance

Reviewer 2: I guess many developers will agree, but evidence appears to be anecdotal. I think this section (and the entire paragraph) would be improved with some references on the topic, although I realize they are hard to come by. I found an (probably outdated?) article on the state of exascale computing in the US (https://doi.org/10.1177%2F1094342014522573) and a paper on the xSDK foundations (https://arxiv.org/pdf/1702.08425.pdf), which both identify (meta-)build systems as essential cornerstones for modern HPC applications. Maybe you could reference them or similar publications?

Author: I was only aware of anecdotal evidence. So I greatly appreciate you finding some actual literature references. I went a head and added those references.

computing, "build system" is increasingly becoming synonymous with CMake
(*N.B.* CMake is often described as a build system generator; however, since
the typical CMake workflow encapsulates running the build system, the

> present manuscript ignores the "generator" distinction). Therefore with the

Reviewer 2: Too specific. This could be moved into a footnote or left out entirely.

Author: I dropped it.

advent of the exascale era, there is a desperate need for a robust CMake
development ecosystem.

CMake already contains a number of tools and features that facilitate
development of the target software package. For example, CMake's
`find_package` module simplifies dependency management, and the `CTest` package
eases the process of testing the resulting software. Additional CMake tools
can be created by writing CMake modules. While the CMake language is
flexible and relatively simple, it is not without its pitfalls. Unfortunately,
tools  to facilitate the development of CMake modules are relatively sparse.
Here we introduce CMinx, a tool for generating API documentation for CMake
modules. API documentation is arguably one of the most basic elements of a
software development ecosystem, and it is our hope that CMinx will serve as the
foundation for a robust CMake development ecosystem.

> Anecdotal evidence [@official_solution] indicates that the developers of the
> CMake language (*i.e.*, Kitware) internally write their API documentation

Reviewer 2: Avoid parentheses
```suggestion
Anecdotal evidence [@official_solution] indicates that Kitware, the developers
of the CMake language, internally write their API documentation
```

using reST and Sphinx. Following best practices, this reST documentation
resides next to the described CMake code. Kitware has also written a
Sphinx plugin that makes it easy to extract the API documentation as part of a
normal Sphinx workflow. This Sphinx plugin is distributed with the source code
for the CMake interpreter and is also available in a GitHub repository mirror
[@sphinx_plugin]. For completeness, we note that similar Sphinx plugins
[@official_sphinx_domain; @marco_koch] have been independently developed but
appear to now be abandoned.

To our knowledge, all of the aforementioned Sphinx plugins simply extract the
reST API documentation verbatim. Notably, this means the developer is
responsible

> for manually: adding the function/macro signatures to the documentation, listing
> the function's parameters, and the overall formatting (*e.g.*, titles and
> subtitles, creation of parameter tables, underline/overline consistency). For

Reviewer 2: Again, the examples are specific in my perspective. Also, the colon appears a bit stray to me.
```suggestion
for manually adding the function/macro signatures to the documentation, listing
the function's parameters, and the overall formatting. For
```

build system developers maintaining one or two CMake modules, these are
admittedly pretty minor inconveniences; however, for organizations
maintaining a substantial CMake code base (such as those for exascale programs),
these "minor inconveniences" can
impact productivity, particularly when ensuring consistency. CMinx
differs from previous solutions primarily in three ways. First, CMinx
understands the grammar of the CMake language, meaning CMinx can automatically
generate some of the documentation by "reading" the source code. Second, CMinx
generates static reST files; this decreases the number of stub files developers
need to maintain and makes it easier for the resulting documentation to be
used in workflows that do not rely on Sphinx. Finally, CMinx has a CMake API
to integrate more easily into existing CMake workflows.

Concurrent with the submission of this manuscript, we have also released the
first production version of CMinx, version 1.0.0. CMinx can be obtained
from the Python Packaging Index via "`pip install CMinx`". Alternatively, CMinx

> can be used as a CMake module via CMake's "`FetchContent`" command.

Reviewer 2: Could you add an example for the `FetchContent` solution to the documentation? I think this would be very helpful, as the CMake modules would become available to users.

Despite only just releasing 1.0.0, the CMinx GitHub organization
has already started to see attention and interest from developers not
affiliated with the authors. So while CMinx might not be considered research
software, we anticipate CMinx to be a useful productivity tool for the large

> swath of research software that uses CMake as their build system.

Reviewer 2: This entire paragraph does not fit into the "Statement of need" and could instead be moved to the end of the "Summary", in my opinion.

Author: Done. I moved it to the end of the summary section.
